### DISCUSSION TITLE
ld: symbol(s) not found for architecture i386

### DISCUSSION THREAD

**Discussion by crispysisig** [rbenv/ruby-build#2351] [Build failures] [answered]
Hello,

I am receiving the following BUILD failure when attempting to install ruby version 2.3.1 via rbenv (`rbenv install 2.3.1`):

Console output:

```
==> Downloading openssl-1.0.2u.tar.gz...
-> curl -q -fL -o openssl-1.0.2u.tar.gz https://dqw8nmjcqpjn7.cloudfront.net/ecd0c6ffb493dd06707d38b14bb4d8c2288bb7033735606569d8f90f89669d16
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 5229k  100 5229k    0     0  18.6M      0 --:--:-- --:--:-- --:--:-- 18.7M
==> Installing openssl-1.0.2u...
-> ./config "--prefix=$HOME/.rbenv/versions/2.3.1/openssl" "--openssldir=$HOME/.rbenv/versions/2.3.1/openssl/ssl" zlib-dynamic no-ssl3 shared no-ssl2 no-krb5
-> make -j 12

BUILD FAILED (macOS 14.3.1 on arm64 using ruby-build 20240116)

You can inspect the build directory at /var/folders/03/m6lw02390glgn2yscbmgbjjc0000gn/T/ruby-build.20240227090202.94502.hCsiXp
See the full build log at /var/folders/03/m6lw02390glgn2yscbmgbjjc0000gn/T/ruby-build.20240227090202.94502.log
```

Log file output:

```
...
ld: symbol(s) not found for architecture i386
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[4]: *** [link_a.darwin] Error 1
make[3]: *** [do_darwin-shared] Error 2
make[2]: *** [libcrypto.1.0.0.dylib] Error 2
make[1]: *** [shared] Error 2
make: *** [build_crypto] Error 1
external command failed with status 2
```

Not sure what `symbol(s) not found for architecture i386` means, so I'm not sure about next steps forward?

**Comment by mislav** [comment]
Hi, Ruby 2.3 is very old (8 years or so) and hasn't been supported by the Ruby core team for already 5 years. You will have trouble compiling this version of Ruby on a modern machine.

If you really really need a Ruby 2.3 installation, you will need to install OpenSSL 1.0, which is also very old and has been unsupported for a while, and doesn't compile on modern systems (as you've already discovered).

[There is a potential workaround](https://github.com/rbenv/homebrew-tap) for installing OpenSSL 1.0 using Homebrew and our unofficial formula, but this is not guaranteed to work.

My strong recommendation for outdated Ruby versions is to stop using them and migrate to a newer, supported version, which will also succeed at compiling on newer machines without much problem.

**Comment by monfresh** [comment, marked as answer]
Like Mislav said, I also highly recommend updating your project to a newer Ruby version, or at a minimum to Ruby 2.3.8. There is rarely a good reason to use a version of Ruby where the patch version (the 3rd digit) is not the latest one. There are known issues with older patch versions, and sometimes simply installing the latest patch version solves the installation problem. Here's a step-by-step guide in case you're not sure how to update your project's Ruby version:

https://www.rubyonmac.dev/how-to-upgrade-the-ruby-version-in-your-project

If you absolutely have to use Ruby 2.3 at this time, and if Mislav's workaround doesn't work for you, the easiest and fastest way to install Ruby 2.3 on macOS is to use my [Ruby on Mac Ultimate](https://www.rubyonmac.dev/) product. It's paid, but 2000+ customers say it's worth it. It's guaranteed to let you install Ruby 2.3.8 (I also just confirmed that 2.3.1 can be installed on macOS 14.3.1). However, if your Ruby project has a Gemfile, I can't guarantee that `bundle install` will work since that will depend on the gems your project depends on and how outdated they are.

**Comment by crispysisig** [comment]
Wonderful, thank you so much for the great answers!
